### PR TITLE
ParseInt Implementation

### PR DESCRIPTION
**Pull request by MaxGraey**
Hi dcode! I'm trying to help you with std functions. 
There is **parseInt** function. I tested this part on pure typescript and it's behave exactly the same as original JS method. For real tests in AS environment need some parser functionality (String literals) which missing yet.

**TODO**
1. Allow parsing binary & octal numeric strings with prefixes 0o, 0b? (But original Javascript method can't do that)
2. Write tests

### CODE REVIEW

**Comment by dcodeIO** [comment]
The tokenizer also has [this one](https://github.com/AssemblyScript/assemblyscript/blob/master/src/tokenizer.ts#L990) derived from tsc. Might make sense to bundle src/util/charcode with stdlib as well so there's a source for the constants.

**Comment by dcodeIO** [comment]
One way to test it, btw., could be to use the charcodes (u16) as the module's memory and then `load<u16>` instead of `charCodeAt` for testing.

**Comment by MaxGraey** [comment]
Got it. May be tomorrow I switch charCodeAt to load<u16> and will write some tests.

**Comment by MaxGraey** [comment]
> One way to test it, btw., could be to use the charcodes (u16) as the module's memory and then load<u16> instead of charCodeAt for testing.

Unfortunately I can't do that because ```ptr``` member in ***String*** is private =(

**Comment by dcodeIO** [comment]
Yeah, `String` cannot be used at all yet. But, if the module's memory would contain the string's u16 char codes starting at offset 0 with its length known, `ptr` would simply be `0` and `length` a constant.

That doesn't lead to something that "just works", of course, it's just the only test case possible atm.

**Comment by MaxGraey** [comment]
Hmm, sounds promising

**Comment by dcodeIO** [comment]
> Allow parsing binary & octal strings? (But original Javascript method can't do that)

Just tested in node.js:

```js
> parseInt("0011", 2)
3
> parseInt("0011", 8)
9
```

**Comment by MaxGraey** [comment]
Current implementation support any radix from 2 to 36.
I mean this misalignment:
```js
> parseInt("0x123")
291
> parseInt("0o10") // expected 8
0
> parseInt("0b10") // expected 2
0
```

**Comment by MaxGraey** [comment]
Hmm, After I add additional file in test directory script ```npm run test:config``` became failure.

**Comment by dcodeIO** [comment]
Yeah, config is tested against tsc, and it doesn't like `std/assembly/string.ts(211,1): error TS1206: Decorators are not valid here.`.

**Comment by MaxGraey** [comment]
> Yeah, config is tested against tsc, and it doesn't like std/assembly/string.ts(211,1): error TS1206: Decorators are not valid here.

So decorators not support for function level yet?

**Comment by dcodeIO** [comment]
> So decorators not support for function level yet?

asc supports decorators everywhere, but tsc doesn't. See my edit above for a work around I have in mind.

**Comment by MaxGraey** [comment]
> Workaround I have in mind currently is to implicitly make all exports from stdlib files globals so functions don't need a @global decorator.

It should& According to meta:

```
FUNCTION_PROTOTYPE: std:string/parseInt
```

But without @global decorator transpiler can't find parseInt in gloabal space, but I try it without test:config

**Comment by MaxGraey** [comment]
The same

```js
ERROR AS2304: Cannot find name 'parseInt'.

 parseInt(changetype<String>(""));
 ~~~~~~~~
 in std/string.ts(3,1)

Error: stack
    at Program.DiagnosticEmitter.emitDiagnostic (/Volumes/Archive/Projects/Github/assemblyscript/src/diagnostics.ts:183:27)
    at Program.DiagnosticEmitter.error (/Volumes/Archive/Projects/Github/assemblyscript/src/diagnostics.ts:188:10)
    at Program.resolveIdentifier (/Volumes/Archive/Projects/Github/assemblyscript/src/program.ts:1011:10)
    at Program.resolveExpression (/Volumes/Archive/Projects/Github/assemblyscript/src/program.ts:1107:21)
    at Compiler.compileCallExpression (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:2337:33)
    at Compiler.compileExpression (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:1108:21)
    at Compiler.compileExpressionStatement (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:887:21)
    at Compiler.compileStatement (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:787:21)
    at Compiler.compileSource (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:316:27)
    at Compiler.compile (/Volumes/Archive/Projects/Github/assemblyscript/src/compiler.ts:219:14)
```

**Comment by MaxGraey** [comment]
Hmm, after sync with base compiler/std/string.ts failure at diff snapshot reason

**Comment by dcodeIO** [comment]
That's expected when replacing var with let because var maps 1:1 to webassembly locals while let, in blocks, is only valid in that block, basically resulting in a shared wasm local.

**Comment by MaxGraey** [comment]
I don't think so problem in ```let``` in this case. But following you description I reverted this changes

**Comment by MaxGraey** [comment]
It seems problem in new changes in std Heap. Can you check?

**Comment by dcodeIO** [comment]
The other differences in the the latest log are expected as well because there have been changes to other parts of the standard library meanwhile. For example the `std:heap/Heap` class doesn't exist anymore in this form, but directly implements `allocate_memory` as functions.

**Comment by MaxGraey** [comment]
So can I somehow fix that?

**Comment by dcodeIO** [comment]
Yes, just regenerate the outdated fixture your code is compared to using the command above.

**Comment by MaxGraey** [comment]
Got it!

File: std/assembly/string.ts [hunk 1]
```diff
@@ -189,8 +200,71 @@ function isWhiteSpaceOrLineTerminator(c: u16): bool {
...
+  var neg = (s0 == cn);
+
+  if (s0 == cp || neg) {
+    if (len == 1)
+      return NaN;
+
+    pos = 1;
+  }
+
+  if (radix == 0) {
```
Comment (dcodeIO) [inline std/assembly/string.ts]: Wouldn't this be sufficiently covered by a `radix: i32 = 10` parameter above?

Comment (MaxGraey) [reply std/assembly/string.ts]: These are fully justified changes. You can try in js:

```js
console.log(parseInt(10)) // => 10
or
console.log(parseInt('10')) // => 10
```

It's clear. But what about specified radix?
```js
console.log(parseInt('0x10', 10)) => 0
console.log(parseInt('0x10')) // => 16
```

So radix with zero value mean actually auto radix

Comment (dcodeIO) [reply std/assembly/string.ts]: I see, makes sense.

**Comment by MaxGraey** [comment]
I added tests for parseInt. Only one case if fail which produce unreachable issue. Need more investigation

File: tests/compiler/std/string.ts [hunk 2]
```diff
@@ -1,6 +1,17 @@
...
 
 var str: string = "hi, I'm a string";
+var str123: string = "123";
+var str123ab: string = "123ab";
+var strab123: string = "ab123";
+var strHex: string = "0x10";
+var strHexInvalid: string = "0x";
+var strHexnInvalid: string = "-0x";
+var strHexnnInv: string = "--0x10";
+// var strHexInvChar: string = "0x1g";
```
Comment (MaxGraey) [inline tests/compiler/std/string.ts]: In this case we get unreachable

Comment (dcodeIO) [reply tests/compiler/std/string.ts]: Hmm, that's strange. Probably hitting an assertion somewhere then. Might make sense to hook `assert` up to an import on failure now, so we know where that is :)

**Comment by dcodeIO** [comment]
To debug this further, you can now import an `abort` function that receives all the information necessary.

```ts
@global
declare function abort(message: string | null, filename: string, line: i32, column: i32): void;

...
```

On the JS side, that'd look like:

```js
function abort(messagePtr, filenamePtr, line, column) {
  ...
}
```

with `messagePtr` and `filenamePtr` pointing at the string in memory, that is a 32-bit length N + N * 16-bit charcodes.

File: tests/compiler/std/string.ts (hunk 2 above)
Comment (MaxGraey) [reply tests/compiler/std/string.ts]: It always print:
```
Testing compiler/std/string.ts
parse incl. I/O: 4.978ms / compile: 4.763ms
validate OK
[trap unreachable]
interpret ERROR:undefined
Created
Created optimized
```

And can't figure out what is problem

**Comment by dcodeIO** [comment]
I created something above that might help as a starting point. It possibly doesn't catch all the corner cases yet and has only rudimentary tests.

**Comment by dcodeIO** [comment]
How do you like to proceed with this one? Was probably a bit insensible of me to implement something on my own though you were working on something already, sorry for that.

**Comment by MaxGraey** [comment]
It's ok, you already implement this so I close this
